Saatvik Sandal 114378631

In [1]:
import pandas as pd
import numpy as np
import matplotlib as mp
import math as math
import matplotlib.pyplot as plt

# Problem 1.1

Table Creation:

In [2]:
avg_temp = [33, 34, 40, 51, 60, 69, 75, 74, 67, 56, 47, 38]
t = []
# populate t table with correct days, using formula give of 16 + 31*(month enumerated)
for i in range(0, 12):
    t.append(16+(31*i))
t = np.array(t)
t

array([ 16,  47,  78, 109, 140, 171, 202, 233, 264, 295, 326, 357])

In [3]:
# vstack stacks rows on top of each other, so calculate calculates the n 1->4 order terms based off t values
# so first it stacks a row of 1s length of 12 (a0), then it stacks the next row of all the t values, (16, 47, 78...), 
# then all the t values again ^2, and again ^3, then it transposes all of it
A = np.vstack([np.ones(len(t)), t, t**2, t**3]).T
#print(A)


# solve normal equation for least squares: (A^T A)x = (A^T)y

#get A^T A
ATA = A.T @ A
# get A^T y
ATY = A.T @ avg_temp

#a(coeffs) = (X^T X)^-1 X^T y
coefficients = np.linalg.inv(ATA) @ ATY

# extract coeffs from variable
a0, a1, a2, a3 = coefficients
print(f"a0 is: {a0}")
print(f"a1 is: {a1}")
print(f"a2 is: {a2}")
print(f"a3 is: {a3}")


a0 is: 25.94178075290347
a1 is: 0.18713946923810454
a2 is: 0.0009576875408841001
a3 is: -3.990504477229028e-06


# Solution to 1.1 (1)
a0 is: 25.94178075290347
a1 is: 0.18713946923810454
a2 is: 0.0009576875408841001
a3 is: -3.990504477229028e-06


In [4]:
# form a function that represents our new polynomial based off the coefficients we just solved for
def polynomial(x, coefficients):
    a0, a1, a2, a3 = coefficients
    return (a0 + (a1*x) + (a2 * (x**2)) + (a3 * (x**3)))


# to get june 4th date, 16 + 31 * 5 gives us june 16. since we need june 4, subtract 12 days from that. 
# 31 * 5 = 155 
# 155 + 16 = 171
# 171 - 12 = 159
print(f"Temperature on june 4th will be: {polynomial(159, coefficients)}")
print(f"Temperature on december 25th will be: {polynomial(366, coefficients)}")


Temperature on june 4th will be: 63.86770803632953
Temperature on december 25th will be: 27.076780223601105


# Solution to 1.1 (2)
Temperature on june 4th will be: 63.86770803632953
Temperature on december 25th will be: 27.076780223601105


In [5]:
# now to calculate when temperature we be 64.89
#there are 372 days in this kind of year
days = []

#loop through all the days, calculate the temperature, and then add day to days if temperature is 64.89 or greater
for i in range(0, 372):
    if((polynomial(i, coefficients) - 64.89) > 0.01):
        days.append(i)

print(days)

#see when days transition temperature
for day in days:
    print(f"Day {day} has a temperature of: {polynomial(day, coefficients)}")

[165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254, 255, 256, 257, 258, 259, 260, 261, 262, 263, 264, 265, 266, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277, 278, 279, 280, 281, 282, 283, 284]
Day 165 has a temperature of: 64.96699155298789
Day 166 has a temperature of: 65.14322185486026
Day 167 has a temperature of: 65.31739298935506
Day 168 has a temperature of: 65.48948101344544
Day 169 has a temperature of: 65.65946198410455
Day 170 has a temperature of: 65.82731195830553
Day 171 has a temperature of: 65.99300699302148
Day 172 has a temperature of: 66.15652314522558
Day 173 has a t

In [ ]:
# use bisection method to find exact points after finding the general areas to search using previous for loop
bisection_iteration_count = 0
def Bisection(a, b, tolerance, coefficients, target_temp=64.89):
    global bisection_iteration_count
    c = 0
    
    while ((b - a) / 2) > tolerance:
        bisection_iteration_count += 1
        #calc midpoint
        c = (b + a) / 2
        
        # calc polymials and use target temp as stopping point
        f_a = polynomial(a, coefficients) - target_temp
        f_c = polynomial(c, coefficients) - target_temp
        
        # check if close enough to 64.89
        if abs(f_c) < tolerance:
            return c
        
        # determine which way to go
        if f_a * f_c < 0:
            b = c
        else:
            a = c
    
    # return final midpoint
    return c


print(f"X root 1 = {Bisection(163, 167, 1e-3, coefficients)}, temperature of : {polynomial(164.5625, coefficients)}")
print(f"X root 2 = {Bisection(282, 285, 1e-3, coefficients)}, temperature of : {polynomial(284.14453125, coefficients)}")

print("Transition from days 164-165 and days 284-285 is when temperature becomes 64.89")

X root 1 = 164.5625, temperature of : 64.88924940158167
X root 2 = 284.14453125, temperature of : 64.89098860678732
Transition from days 164-165 and days 284-285 is when temperature becomes 64.89


# Solution to 1.1 (3)
X root 1 = 164.5625, temperature of : 64.88924940158167
X root 2 = 284.14453125, temperature of : 64.89098860678732
Transition from days 164-165 and days 284-285 is when temperature becomes 64.89